componentLibraries/extensionLibrary/Connectivity/FMIWrapper.hpp

```diff
@@ -0,0 +1,355 @@
+{
+    hopsan::Component* pComponent = (hopsan::Component*)pComponentEnvironment;
+    if (pComponent == nullptr) {
+        return;
+    }
+
+    char buffer[512];
+    va_list args;
+    va_start(args, message);
+    vsnprintf(buffer, 512, message, args);
+    va_end(args);
+
+    switch (status) {
+    // Typically info messages are not something we want to see in Hopsan, so show them as debug type
+    case fmi2_status_ok:
+        pComponent->addDebugMessage(buffer);
+        break;
+    case fmi2_status_warning:
+        pComponent->addWarningMessage(buffer);
+        break;
+    case fmi2_status_error:
+        pComponent->addErrorMessage(buffer);
+        break;
+    case fmi2_status_fatal:
+        pComponent->addFatalMessage(buffer);
+        break;
+    default:
+        // Discard
+        break;
+    }
+}
+
+
+
+class FMIWrapper : public ComponentSignal
+{
+private:
+    HString mFmuPath, mLastFmuPath, mTmpPath;
+    std::map<fmi2_value_reference_t,double*> mOutputs;
+    std::map<fmi2_value_reference_t,double*> mInputs;
+    std::map<fmi2_value_reference_t,double> mParameters;
+    std::vector<Port*> mPorts;
+
+    jm_callbacks callbacks;
+    jm_status_enu_t status;
+    fmi_import_context_t* context;
+    fmi_version_enu_t version;
+    fmi2_callback_functions_t fmiCallbackFunctions;
+    fmi2_status_t fmistatus;
+    fmi2_import_t* fmu;
+
+
+public:
+    static Component *Creator()
+    {
+        return new FMIWrapper();
+    }
+
+    void configure()
+    {
+        addConstant("path", "Path to functional mockup unit (FMU)", "", mFmuPath);
+    }
+
+    void reconfigure()
+    {
+        if(mFmuPath == mLastFmuPath) {
+            return; //Path did not change, do nothing
+        }
+        mLastFmuPath = mFmuPath;
+
+        deconfigure(); //Make sure to unload FMU and free memory before loading a new one
+
+        for(const auto *port : mPorts) {
+            removePort(port->getName());
+        }
+        std::vector<HString> parameters;
+        this->getParameterNames(parameters);
+        for(const auto &parameter : parameters) {
+            if(parameter != "path") {
+                this->unRegisterParameter(parameter);
+            }
+        }
+        mPorts.clear();
+        mOutputs.clear();
+        mInputs.clear();
+        mParameters.clear();
+
+        callbacks.malloc = malloc;
+        callbacks.calloc = calloc;
+        callbacks.realloc = realloc;
+        callbacks.free = free;
+        callbacks.logger = jmLogger;
+        callbacks.log_level = jm_log_level_debug;
+        callbacks.context = static_cast<jm_voidp>(this);
+
+        context = fmi_import_allocate_context(&callbacks);
+
+        addInfoMessage("Loading FMU from "+mFmuPath+"...");
+
+        //Generate random temporary path
+        mTmpPath = std::filesystem::temp_directory_path().c_str();
```

Comment:
Since only C++17 is used to crate and remove a temp directory, maybe it would not be so difficult to use "native" C code for that instead. Possibly a new utility in HopsanCore that wraps such functions, (it could internally use ifdefs to check for c++17).
I actually recently added "dirent" a MSVC implementation third party header-only library, that is used in the RemoteHopsan libraries to enable dirent based filesystem access also with MSVC. 
Of course lookup of temp directory must be made with native calls.

Not sure whats easier, to add ifdefs with native code (Windows, Posix and MacOS(might work with posix)) directly. Or add a HopsanCore utility which includes dirent for MSVC.

Reply:
FMI import always requres a temp directory, where the FMU is unzipped by FMILibrary. The previous FMI import generates and compiles a new component for each imported FMU. The temp folder is then created directly in the folder for that component library. In the new version a generic component is used for all FMUs. Therefore every component instance needs to create its own temp directory.

We can probably use C++03 to create the temp folder, only drawback is that it requires far more code. The only problem is that Linux systems have no API function for getting the temp directory. The existing functions in C++17 and Boost checks some envirnoment variables (like "TEMPDIR", "TEMPPATH", "TEMP" and so on), and fallbacks to "/tmp" if nothing else is found. Perhaps we can just always assume "/tmp" on Linux systems.

Reply:
I added a utility class for temp folders. It creates a temp folder in constructor and removes in in the destructor, using C++17 if available, otherwise C++03. It has only two member functions, one for fetching the path and one for checking if the temp directory is valid.

I first thought about having a general file manipulation utility. But then I realized that a function for `deleteFolderAndAllSubfoldersRecursively` is quite of dangerous. The new utility class will only delete the same folder it created, which is much safer.

Reply:
yes that would have been unsafe, did you check the env variables, before assuming /tmp ?

Reply:
No, I'm not sure if it was necessary. I did some research on this, and some people suggest only assuming "/tmp" is enough. I guess the folder always exists on Linux systems, and the only reason to use the environment variables is if you really want to put temp files somewhere else. But if you think it is needed I can add it?